jmclient/jmclient/wallet_rpc.py

```diff
@@ -1023,3 +1047,111 @@ def getseed(self, request, walletname):
+
+            # -- Options parsing -----------------------------------------------
+
+            (options, args) = get_tumbler_parser().parse_args([])
+            self.tumbler_options = vars(options)
+
+            # At the moment only the destination addresses can be set.
+            # For now, all other options are hardcoded to the defaults of
+            # the tumbler.py script.
+            request_data = self.get_POST_body(request, ["destination_addresses"])
+            if not request_data:
+                raise InvalidRequestFormat()
+
+            destaddrs = request_data["destination_addresses"]
+            for daddr in destaddrs:
+                success, _ = validate_address(daddr)
+                if not success:
+                    raise InvalidRequestFormat()
+
+            # Setting max_cj_fee based on global config.
+            # We won't respect it being set via tumbler_options for now.
+            def dummy_user_callback(rel, abs):
+                raise ConfigNotPresent()
+
+            max_cj_fee = get_max_cj_fee_values(jm_single().config,
+                                               None,
+                                               user_callback=dummy_user_callback)
+
+            jm_single().mincjamount = self.tumbler_options['mincjamount']
+
+            # -- Schedule generation -------------------------------------------
+
+            # Always generates a new schedule. No restart support for now.
+            schedule = get_tumble_schedule(self.tumbler_options,
+                                           destaddrs,
+                                           self.services["wallet"].get_balance_by_mixdepth())
+
+            logsdir = os.path.join(os.path.dirname(jm_single().config_location),
+                                   "logs")
+            sfile = os.path.join(logsdir, self.tumbler_options['schedulefile'])
+            with open(sfile, "wb") as f:
+                f.write(schedule_to_text(schedule))
+
+            tumble_log = get_tumble_log(logsdir)
+            tumble_log.info("TUMBLE STARTING")
+            tumble_log.info("With this schedule: ")
+            tumble_log.info(pprint.pformat(schedule))
+
+            # -- Running the Taker ---------------------------------------------
+
+            # For simplicity, we're not doing any fee estimation for now.
+            # We might want to add fee estimation (see scripts/tumbler.py) to
+            # prevent users from overspending on fees when tumbling with small
+            # amounts.
+
+            if not self.activate_coinjoin_state(CJ_TAKER_RUNNING):
+                raise ServiceAlreadyStarted()
+
+            self.taker = Taker(self.services["wallet"],
+                               schedule,
+                               max_cj_fee=max_cj_fee,
+                               order_chooser=self.tumbler_options['order_choose_fn'],
+                               callbacks=(self.filter_orders_callback_tumbler, None,
```

Comment:
The second callback (the `taker_info_callback`) is `None` for the CLI because the update messages are printed out to log (`jmclient.Taker.default_taker_info_callback`). One reason we really need that kind of info update callback is to show the user in case the tumbler got stalled for a lack of PoDLE commitments, which *can* mean the tumbler is *permanently* stalled (if there are no utxos which have unused commitments). In the JoinmarketQt app we use that info callback to send back that info.

We *should* do the same here, this will be one of 2 or 3 reasons why a user might find their tumbler just stopped doing anything, and it can be kind of tough on users (to say the least!) because by default it *does* wait, sometimes for hours, including long confirm times and deliberate long delays, and so unless they're getting the kind of rich feedback that you see in `logs/TUMBLE.log` and `logs/TUMBLE.schedule` plus actually knowing how to read the schedules, they're often really pretty in the dark.

I think this really just illustrates that there's quite a *lot* of work to move this from "MVP" to something not so "M" :)

Reply:
Very good points! Ideally we'd be able to stream this feedback directly to the web client (via web-sockets). I'll try to see if I can get something like that working -- ideally in a separate PR after this one is merged, if that's ok for you.